packages/langium/src/default-module.ts

```diff
@@ -30,35 +30,58 @@ import { DefaultNameProvider } from './references/naming';
 import { DefaultReferences } from './references/references';
 import { DefaultScopeComputation, DefaultScopeProvider } from './references/scope';
 import { DefaultJsonSerializer } from './serializer/json-serializer';
-import { LangiumDefaultServices, LangiumServices } from './services';
+import { DeepPartial, LangiumDefaultServices, LangiumDefaultSharedServices, LangiumGeneratedServices, LangiumGeneratedSharedServices, LangiumServices, LangiumSharedServices } from './services';
 import { DefaultDocumentValidator } from './validation/document-validator';
 import { ValidationRegistry } from './validation/validation-registry';
+import { ExtensionServiceRegistry, ServiceRegistry, SingleServiceRegistry } from './service-registry';
 
-export type DefaultModuleContext = {
+export interface DefaultModuleContext {
     connection?: Connection
+    serviceRegistry?: ServiceRegistry
 }
 
-export function createDefaultModule(context: DefaultModuleContext = {}): Module<LangiumServices, LangiumDefaultServices> {
+export function injectSingleService<T extends LangiumServices>(
+    sharedModule: Module<LangiumSharedServices, LangiumDefaultSharedServices>,
+    generatedSharedModule: Module<LangiumSharedServices, LangiumGeneratedSharedServices>,
+    generatedModule: Module<LangiumServices, LangiumGeneratedServices>,
+    module?: Module<T, DeepPartial<T>>): T {
+
+    const shared = inject(sharedModule, generatedSharedModule);
+    const single = inject(createDefaultModule(shared), generatedModule, module);
+    shared.ServiceRegistry = new SingleServiceRegistry(single);
+    return single;
+}
+
+export function createSharedModule(context: DefaultModuleContext = {}): Module<LangiumSharedServices, LangiumDefaultSharedServices> {
     return {
-        parser: {
-            GrammarConfig: (injector) => createGrammarConfig(injector),
-            LangiumParser: (injector) => createLangiumParser(injector),
-            ValueConverter: () => new DefaultValueConverter(),
-            TokenBuilder: () => new DefaultTokenBuilder()
+        ServiceRegistry: () => context.serviceRegistry ?? new ExtensionServiceRegistry(),
+        lsp: {
+            Connection: () => context.connection
         },
-        documents: {
+        workspace: {
             LangiumDocuments: (injector) => new DefaultLangiumDocuments(injector),
             LangiumDocumentFactory: (injector) => new DefaultLangiumDocumentFactory(injector),
             DocumentBuilder: (injector) => new DefaultDocumentBuilder(injector),
             TextDocuments: () => new TextDocuments(TextDocument),
             TextDocumentFactory: (injector) => new DefaultTextDocumentFactory(injector),
+            IndexManager: (injector) => new DefaultIndexManager(injector)
+        }
+    };
+}
+
+export function createDefaultModule(shared: LangiumSharedServices): Module<LangiumServices, LangiumDefaultServices> {
```

Comment:
Idea for simplifying the DI concept: Instead of generating separate modules for `LangiumGeneratedServices`, provide them as a simple object to be passed as additional parameter to `createDefaultModule`. Then they are included in the module returned here. Same approach for `LangiumGeneratedSharedServices` and `createSharedModule`.

It's only a small improvement, but would reduce the number of "services" and "module" types.